particle_structs/src/ps_for.hpp

```diff
@@ -53,4 +53,29 @@ namespace pumipic {
     throw 1;
     return NULL;
   }
+
+  //This function initializes and populates the pids and offsets arrays
+  template <typename DataTypes, typename MemSpace>
+  template <typename ViewT>
+  void ParticleStructure<DataTypes, MemSpace>::getPIDs(ViewT& pids, ViewT& offsets) {
+    offsets = ViewT("offsets", num_elems+1);
+    ViewT ppe("ppe", num_elems+1);
+    auto setPPE = PS_LAMBDA(const lid_t& e, const lid_t& p, const bool& mask) {
+      if (mask) {
+        Kokkos::atomic_increment(&ppe(e));
+      }
+    };
+    parallel_for(this, setPPE, "setPPE");
+    exclusive_scan(ppe, offsets, execution_space());
+    
+    pids = ViewT("pids", getLastValue(offsets));
+    ViewT currIndex("currIndex", num_elems);
+    auto setPIDs = PS_LAMBDA(const lid_t& e, const lid_t& p, const bool& mask) {
+      if (mask) {
+        auto index = Kokkos::atomic_fetch_add(&currIndex(e), 1);
+        pids(offsets(e)+index) = p;
+      }
+    };
+    parallel_for(this, setPIDs, "setPIDs");
+  }
```

Comment:
@cwsmith Thank you for the input on the previous function. This version groups the pids by element and they are packed together. There is a side effect that the order of pids is not preserved. Would this be an issue? If it looks good to you then I will write up a test for this function.

Reply:
You're welcome.  I asked Chonglin about the order in the issue for this PR.